10up/ElasticPress
# Support customizing the template for search results in Instant Results

### Description of the Change
This implements a JavaScript hook using `@wordpress/hooks` to support developers replacing the component used to render search results in Instant Results.

In #2527 I described an approach where the filter wraps the return value to filter the rendered element upon render, but in this PR I ended up taking a different approach where I am filtering the component itself. I did this for a few reasons:

1. The filter only runs once, when the component is defined.
2. Conditionally replacing the component per-result is still possible via a wrapper component approach, which I've documented.
3. I think it's easier to understand what needs to be returned from any filter callbacks, as a React component is a well known concept, whereas filtering just the rendered result could cause confusion and give the incorrect impression that raw HTML could be returned to the filter or that other rendering methods would be supported.

Something that was also considered was the naming of JavaScript hooks. @felipeelia and I explored a few other plugins and Gutenberg itself and didn't find much consistency. What this PR proposes is `plugin.Feature.valueBeingFiltered` In this case that's `elasticpress.InstantResults.Result`.

The `Result` value is capitalized to reflect the fact that a React component is being filtered and those are conventionally named, and represented in JSX, capitalized. If the value of a regular function such as `getPostTypesFromForm()` were to be filtered that would use `elasticpress.InstantResults.getPostTypesFromForm`.

Some other things to think about are whether the namespace should be `elasticpress` or `ep`, to be somewhat consistent with the PHP hooks, and how to capitalize the feature names since they don't represent an actual object in JavaScript. I went with `InstantResults` over `instantResults` to match the PHP class and for aesthetic reasons.


Closes #2527

### How to test the Change
The included documentation update includes some examples of how the filter can be used. Implementing these would confirm if things are functioning as expected.

### Changelog Entry
> Added - `elasticpress.InstantResults.Result` JavaScript filter for filtering the component used for Instant Results search results.

### Credits
Props @JakePT 

### Checklist:
